test/infrastructure/docker/exp/internal/controllers/dockermachinepool_controller.go

```diff
@@ -198,30 +237,32 @@ func (r *DockerMachinePoolReconciler) reconcileNormal(ctx context.Context, clust
 		machinePool.Spec.Replicas = pointer.Int32(1)
 	}
 
-	pool, err := docker.NewNodePool(ctx, r.Client, cluster, machinePool, dockerMachinePool)
-	if err != nil {
-		return ctrl.Result{}, errors.Wrap(err, "failed to build new node pool")
+	if err := r.createNewInstances(ctx, cluster, machinePool, dockerMachinePool); err != nil {
+		return ctrl.Result{}, err
 	}
 
-	// Reconcile machines and updates Status.Instances
-	remoteClient, err := r.Tracker.GetClient(ctx, client.ObjectKeyFromObject(cluster))
-	if err != nil {
-		return ctrl.Result{}, errors.Wrap(err, "failed to generate workload cluster client")
+	if err := r.deleteExtraDockerMachines(ctx, cluster, machinePool, dockerMachinePool); err != nil {
+		return ctrl.Result{}, err
 	}
-	res, err := pool.ReconcileMachines(ctx, remoteClient)
+
+	if err := r.createMissingDockerMachines(ctx, cluster, machinePool, dockerMachinePool); err != nil {
```

Comment:
In `createNewInstances()`, we attempt to create a DockerMachine after creating each container as follows. If `createDockerMachine()` fails, we return an error, and we want to make sure the DockerMachine gets created on the next Reconcile loop. 

```
numToCreate := int(*machinePool.Spec.Replicas) - matchingMachineCount
for i := 0; i < numToCreate; i++ {
	log.V(2).Info("Creating a new Docker container for machinePool", "machinePool", machinePool.Name)
	createdMachine, err := createDockerContainer(ctx, cluster, machinePool, dockerMachinePool)
	if err != nil {
		return errors.Wrap(err, "failed to create a new docker machine")
	}

	log.V(2).Info("Creating a new DockerMachine for Docker container", "container", createdMachine.Name())
	// Note: we want the DockerMachine to have the same name as the container so that we can easily find it later.
	if err := r.createDockerMachine(ctx, createdMachine.Name(), cluster, machinePool, dockerMachinePool); err != nil {
		return errors.Wrap(err, "failed to create a new docker machine")
	}
}
```

Reply:
based on my experience, when re-entrancy for an error in a part of the codebase is implemented in another part of the codebase, there are high chances that we will lose track of this and get regression down the line 😞 

Unfortunately, I don't know when I can have a dedicated slot of time to think of better solutions, but for sure there are. Please, if you can avoid this, or at least invest more in the comments so those cross dependencies between parts of the codebase are well tracked + add a specific unit test

Reply:
Sounds good, let me discuss with @CecileRobertMichon to see if we can come up with an alternative.